ext/js/language/translator.js

```diff
@@ -236,49 +248,153 @@ export class Translator {
+        /** @type {import('translation-internal').DatabaseDeinflection[]} */
+        const dictionaryDeinflections = [];
+        deinflections.forEach((deinflection) => {
+            const {originalText, transformedText, inflectionHypotheses: algHypotheses, databaseEntries} = deinflection;
+            databaseEntries.forEach((entry) => {
+                const {dictionary, definitions} = entry;
+                const entryDictionary = enabledDictionaryMap.get(dictionary);
+                const useDeinflections = entryDictionary?.useDeinflections ?? true;
+                if (!useDeinflections) { return; }
+                for (const definition of definitions) {
+                    if (Array.isArray(definition)) {
+                        const [formOf, inflections] = definition;
+                        if (!formOf) { continue; }
+
+                        const inflectionHypotheses = algHypotheses.map(({inflections: algInflections}) => {
+                            return {
+                                source: /** @type {import('dictionary').InflectionSource} */ (algInflections.length === 0 ? 'dictionary' : 'both'),
+                                inflections: [...algInflections, ...inflections]
+                            };
+                        });
+
+                        const dictionaryDeinflection = this._createDeinflection(originalText, transformedText, formOf, 0, inflectionHypotheses);
+                        dictionaryDeinflections.push(dictionaryDeinflection);
+                    }
+                }
+            });
+        });
+
+        await this._addEntriesToDeinflections(dictionaryDeinflections, enabledDictionaryMap, matchType, false);
+
+        return dictionaryDeinflections;
+    }
+
+    /**
+     * @param {import('translation-internal').DatabaseDeinflection[]} deinflections
+     * @param {Map<string, import('translation').FindTermDictionary>} enabledDictionaryMap
+     * @param {import('dictionary').TermSourceMatchType} matchType
+     * @param {boolean} partsOfSpeechFilter
+     */
+    async _addEntriesToDeinflections(deinflections, enabledDictionaryMap, matchType, partsOfSpeechFilter = true) {
+        const uniqueDeinflectionsMap = this._groupDeinflectionsByTerm(deinflections);
+        const uniqueDeinflectionArrays = Array.from(uniqueDeinflectionsMap.values());
+        const uniqueDeinflectionTerms = Array.from(uniqueDeinflectionsMap.keys());
+
         const databaseEntries = await this._database.findTermsBulk(uniqueDeinflectionTerms, enabledDictionaryMap, matchType);
+        this._matchEntriesToDeinflections(databaseEntries, uniqueDeinflectionArrays, partsOfSpeechFilter);
+    }
+
+    /**
+     * @param {import('translation-internal').DatabaseDeinflection[]} deinflections
+     * @returns {Map<string, import('translation-internal').DatabaseDeinflection[]>}
+     */
+    _groupDeinflectionsByTerm(deinflections) {
+        const result = deinflections.reduce((map, deinflection) => {
```

Comment:
IMO this is more clear and consistent with other parts of the codebase if you don't use reduce.

```js
/** @type {Map<string, import('translation-internal').DatabaseDeinflection[]>} */
const result = new Map();
for (const deinflection of deinflections) {
    const {deinflectedText} = deinflection;
    let deinflectionArray = result.get(deinflectedText);
    if (typeof deinflectionArray === 'undefined') {
        deinflectionArray = [];
        result.set(deinflectedText, deinflectionArray);
    }
    deinflectionArray.push(deinflection);
}
return result;
```
(this impl also avoids redundant `.set` which is also more consistent with other parts of translator)